utils/land-patterns.stanza

```diff
@@ -877,31 +877,52 @@ public pcb-landpattern dfn-landpattern ( n-pad:Int, pad-pitch:Double, pad-dims:[
   make-dfn-landpattern(n-pad,pad-pitch,pad-dims,pad-x0,body-dims)
   pad p[n-pad + 1] : smd-pad(hs[0],hs[1]) at loc(0.0, 0.0)  
 
-; returns the letter corresponding to a row in a bga package, for example:
-; A, B, C, D, E, E, F, G, H, J, K, L, M, N, P, R, T, U, V, W, Y
-; AA, AB, ... etc
-public defn red-alph-letter (idx:Int):
-  val table = "ABCDEFGHJKLMNPRTUVWY"
+; Returns the name of a row given its index and a list of omitted
+; characters.
+; 
+; The algorithm is a realization of this description from JEDEC JC-11:
+;
+; > When a package is viewed ... with the A1 corner in the upper left position, 
+; > terminal rows are lettered from top to bottom starting with A and 
+; > terminal columns are numbered from left to right starting with 1. 
+; > ... For packages with greater than 20 rows (i.e.,  following rows W and Y),  
+; > the 21st row is designated AA and subsequent rows are designated AB, AC, etc.
+;
+; The modification is we abstract over the omitted characters to support non-
+; standard naming conventions. 
+;
+public defn bga-row-name (idx:Int, omitted:String) -> Ref:
+  val table = to-tuple $
+    for ch in "ABCDEFGHIJKLMNOPQRSTUVWXYZ" filter:
+      not contains?(upper-case(omitted), ch)
   val buf = StringBuffer()
   let loop (row: Int = idx):
     if row >= length(table):
       print(buf, table[row / length(table) - 1])
       loop(row / length(table))
   print(buf, table[idx % length(table)])
-  to-string(buf)
+  Ref(to-string(buf))
 
-public defn bga-pin-name (r:Int, c:Int) :
-  val letter = red-alph-letter(r)
-  Ref(string-join([letter, c + 1]))
+; for backwards compat
+public defn red-alph-letter (r:Int):
+  to-string(bga-row-name(r, "IOQSXZ"))
+
+public defn bga-pin-name (row:Int, column:Int, omitted?:String|False) -> Ref:
+  val omitted = match(omitted?):
+    (_:False): "IOQSXZ"
+    (s:String): s
+  bga-row-name(row, omitted)[column + 1]
+
+public defn bga-pin-name (row:Int, column:Int) -> Ref:
+  bga-pin-name(row, column, false)
 
 public defn make-bga-pkg (pitch:Double, pad-diam:Double, n-pads:[Int Int], courtyard:[Double Double], omit-pads:Tuple<Ref>|False) :
   inside pcb-landpattern :
     val pin-pad = bga-pad(pad-diam)
     defn make-pad (r:Int, c:Int) :
       val x = ((to-double(n-pads[0]) - 1.0) / -2.0 + to-double(c)) * pitch
       val y = ((to-double(n-pads[1]) - 1.0) / 2.0 - to-double(r)) * pitch
-      val letter = red-alph-letter(r)
-      val name = Ref(letter)[c + 1]
+      val name = bga-pin-name(r, c)
```

Comment:
For performance-sake, can we change this call to something like:

```
val pin-pad = bga-pad(pad-diam)
val pad-names = bga-pad-names(...)
defn make-pad (r:Int, c:Int) : 
  val x = ...
  val y = ...
  val name = pad-names[r,c]
  ...
```

Reply:
Is `pad-names[r,c]` valid syntax? I'm not sure what that means or what the return type of `bga-pad-names` should be given the context (`Tuple<Tuple<Ref>>`?)

Reply:
I am thinking it's some object that supports a `get` method. 

e.g. 
```
```

Reply:
Ok, should be resolved now. I made the same change in the make-bga-landpattern generator, and precomputed the pad names.